### PR TITLE
Drop clause if it contains an impossible BinaryConstraint

### PR DESCRIPTION
**Pull request by adamjseitz**
Some BinaryConstraint atoms can be determined to be impossible during synthesis. This PR introduces a transform to eliminate clauses containing such impossible atoms.

These kinds of atoms are not likely to be written directly, but can easily arise with inlined code.

In ddisasm (with a small work-in-progress change to take greater advantage of this), across its various arch configurations, Souffle generates between 5.0% and 12% fewer lines of C++ with this patch. This yields build time improvements of 2.5% - 10% depending on arch configuration and compiler.

This is my first attempt at adding a transformer pass, so please apply additional scrutiny.

### CODE REVIEW

**Review by quentin — COMMENTED** [review]
Could this be done more generally by introducing a constant folding transformer that runs before the `RemoveBooleanConstraintsTransformer` ?

The `RemoveBooleanConstraintsTransformer` removes clauses that contain the `false` constraint.

The constant folding transformer would transform non-satisfiable binary constraints into the `false` `BinaryConstraint`, and always-satisfiable binary constraints into the `true` `BinaryConstraint`.

Other constant folding operations could be performed as well.

**Comment by adamjseitz** [comment]
> Could this be done more generally by introducing a constant folding transformer that runs before the `RemoveBooleanConstraintsTransformer` ?
> 
> The `RemoveBooleanConstraintsTransformer` removes clauses that contain the `false` constraint.
> 
> The constant folding transformer would transform non-satisfiable binary constraints into the `false` `BinaryConstraint`, and always-satisfiable binary constraints into the `true` `BinaryConstraint`.
> 
> Other constant folding operations could be performed as well.

That sounds to me like an approach that would make sense, although we'd have to do some tweaking of the transform pipeline order to get everything working as desired. As I noted, these kind of impossible constraints arise most often in inlined relations. I was only able to get my changes to have the desired result with it after both `InlineRelationsTransformer` and `ResolveAliasesTransformer` in the pipeline. `RemoveBooleanConstraintsTransformer` currently runs before inlining in the pipeline. 

Do you think it would make sense to move `RemoveBooleanConstraintsTransformer` until after inlining in the pipeline? Or should it be run again, similar to `ResolveAliasesTransformer`, which runs several times?

**Comment by adamjseitz** [comment]
Hi @quentin,

I pushed another commit to refactor the new transform as `SimplifyBinaryConstraintsTransform`, which just rewrites `BinaryConstraint` into `BooleanConstrant`, and added another pass of `RemoveBooleanConstraintsTransformer` after it runs.

Of course, we could support rewriting more `BinaryConstraint` operand types, but for now I've just left it supporting `EQ` and `NE` .

Let me know if you have further feedback!

File: src/ast/transform/SimplifyConstantBinaryConstraints.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+#pragma once
+
+#include "ast/TranslationUnit.h"
+#include "ast/transform/Transformer.h"
+#include <string>
+
+namespace souffle::ast::transform {
+
+/**
+ * Eliminate clauses containing unsatisfiable constant binary constraints
```
Comment (quentin) [inline src/ast/transform/SimplifyConstantBinaryConstraints.h:24]: this transform does not eliminate clauses, it simplifies binary constraints into boolean constraints.

**Review by quentin — CHANGES_REQUESTED** [review]

File: src/ast/transform/SimplifyConstantBinaryConstraints.h (hunk 1 above)
Comment (adamjseitz) [reply src/ast/transform/SimplifyConstantBinaryConstraints.h]: Good catch, updated the docstring.

**Review by quentin — APPROVED** [review]

**Pull request merged**